src/types/index.ts

```diff
@@ -172,7 +197,11 @@ export interface SubscriptionIOS extends ProductCommon {
   subscriptionPeriodUnitIOS?: SubscriptionIosPeriod;
 }
 
-export type Subscription = SubscriptionAndroid & SubscriptionIOS;
+export type Subscription =
```

Comment:
@hsource The downside I saw to using `|` instead of `&` is that then you'll have to check the type. I had the idea of eventually converting the type to something more like:
```
export type Subscription = {
android: SubscriptionAndroid,
amazon:SubscriptionAmazon,
ios: SubscriptionIOS,
}
```
I'm very curious to read your thoughts

Reply:
@andresesfm 

I think the main benefit of this is that `Subscription` will only ever be one of `SubscriptionAndroid` or `SubscriptionIOS`. This lets users more easily have `if` and `switch` statements that work for both iOS and Android.

The documentation on this is here: [discriminated unions](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#discriminated-unions)

Here's an example in my real code code:

```ts
function getLocalizedPriceFromSubscription(subscription: Subscription) {
  switch (subscription.platform) {
    case SubscriptionPlatform.android: {
      // Typescript uses discriminated unions to know that the subscription is a SubscriptionAndroid
      const pricingPhaseInfo = getPaidPricingPhaseInfo(subscription);
      return pricingPhaseInfo.formattedPrice;
    }
    case SubscriptionPlatform.ios:
      // Typescript uses discriminated unions to know that the subscription is a SubscriptionIos
      return subscription.localizedPrice;
    case SubscriptionPlatform.amazon:
      throw new Error(
        'getLocalizedPriceFromSubscription: Amazon is currently not supported',
      );
  }
}
```

This makes it much easier to work with the subscription types. This way, we can mark less properties as optional too, which makes working with the types easier.

Reply:
If we converted the type to this:

```ts
export type Subscription = {
  android?: SubscriptionAndroid,
  amazon?: SubscriptionAmazon,
  ios?: SubscriptionIOS,
}
```

The user would still also have to check the type by doing `if (subscription.android) { ... } else if (subscription.ios) { ... }`

I think using union types is a bit clearer because it makes it clear that only one of the 3 will be filled at a time.

Reply:
If we converted the type to this:

```ts
export type Subscription = {
  android?: SubscriptionAndroid,
  amazon?: SubscriptionAmazon,
  ios?: SubscriptionIOS,
}
```

The user would still also have to check the type by doing `if (subscription.android) { ... } else if (subscription.ios) { ... }`

I think using union types is a bit clearer because it makes it clear that only one of the 3 will be filled at a time.